domain/schema/secret.go

```diff
@@ -309,25 +309,86 @@ CREATE TABLE
 CREATE UNIQUE INDEX idx_secret_role_role ON secret_role (role);
 
 INSERT INTO secret_role VALUES
-    (0, 'view'),
-    (1, 'rotate'),
+    (0, 'none'),
+    (1, 'view'),
     (2, 'manage');
 
--- TODO: probably we need a model level permission table like the
--- recently added "permission" table in the controller database.
+CREATE TABLE
+    secret_grant_subject_type (
+        id   INT PRIMARY KEY,
+        type TEXT
+    );
+
+INSERT INTO secret_grant_subject_type VALUES
+    (0, 'unit'),
+    (1, 'application'),
+    (2, 'model');
+
+CREATE TABLE
+    secret_grant_scope_type (
+        id   INT PRIMARY KEY,
+        type TEXT
+    );
+
+INSERT INTO secret_grant_scope_type VALUES
+    (0, 'unit'),
+    (1, 'application'),
+    (2, 'model'),
+    (3, 'relation');
+
 CREATE TABLE
     secret_permission (
-        uuid TEXT PRIMARY KEY,
-        scope TEXT NOT NULL,
-        subject TEXT NOT NULL,
-        role TEXT NOT NULL,
-        CONSTRAINT chk_empty_scope
-            CHECK(scope != ''),
-        CONSTRAINT chk_empty_subject
-            CHECK(subject != ''),
+        secret_id TEXT NOT NULL,
+        role_id INT NOT NULL,
+        -- subject_uuid is the entity which  
+        -- has been granted access to a secret.  
+        -- It will be an application, unit, or model uuid.  
+        subject_uuid TEXT NOT NULL,
+        subject_type_id INT NOT NULL,
+        -- scope_uuid is the entity which  
+        -- defines the scope of the grant.  
+        -- It will be an application, unit, relation, or model uuid.  
+        scope_uuid TEXT NOT NULL,
+        scope_type_id TEXT NOT NULL,
+        CONSTRAINT pk_secret_permission_secret_id_subject_uuid
+            PRIMARY KEY (secret_id,subject_uuid),
+        CONSTRAINT chk_empty_scope_uuid
+            CHECK(scope_uuid != ''),
+        CONSTRAINT chk_empty_subject_uuid
+            CHECK(subject_uuid != ''),
+        CONSTRAINT fk_secret_permission_secret_id
+            FOREIGN KEY (secret_id)
+            REFERENCES secret_metadata (secret_id),
         CONSTRAINT fk_secret_permission_secret_role_id
-            FOREIGN KEY (role)
-            REFERENCES secret_role (role)
+            FOREIGN KEY (role_id)
+            REFERENCES secret_role (id),
+        CONSTRAINT fk_secret_permission_secret_grant_subject_type_id
+            FOREIGN KEY (subject_type_id)
+            REFERENCES secret_grant_subject_type (id),
+        CONSTRAINT fk_secret_permission_secret_grant_scope_type_id
+            FOREIGN KEY (scope_type_id)
+            REFERENCES secret_grant_scope_type (id)
     );
+
+CREATE INDEX idx_secret_permission_secret_id ON secret_permission (secret_id);
+CREATE INDEX idx_secret_permission_subject_uuid_subject_type_id ON secret_permission (subject_uuid,subject_type_id);
+
+CREATE VIEW v_secret_permission AS
```

Comment:
I think this can be done more efficiently.
```
ID      PARENT  NOTUSED DETAIL

2       0       0       SCAN sp
11      0       0       CORRELATED SCALAR SUBQUERY 3
16      11      0       SEARCH unit USING INDEX sqlite_autoindex_unit_1 (uuid=?)
29      0       0       CORRELATED SCALAR SUBQUERY 4
34      29      0       SEARCH application USING INDEX sqlite_autoindex_application_1 (uuid=?)
48      0       0       SCALAR SUBQUERY 5
52      48      0       SCAN model USING COVERING INDEX sqlite_autoindex_model_1
65      0       0       CORRELATED SCALAR SUBQUERY 6
70      65      0       SEARCH unit USING INDEX sqlite_autoindex_unit_1 (uuid=?)
83      0       0       CORRELATED SCALAR SUBQUERY 7
88      83      0       SEARCH application USING INDEX sqlite_autoindex_application_1 (uuid=?)
102     0       0       SCALAR SUBQUERY 8
106     102     0       SCAN model USING COVERING INDEX sqlite_autoindex_model_1
```

Reply:
Optimised it
ID      PARENT  NOTUSED DETAIL                                                                  

11      0       0       SCAN sp                                                                 
13      0       0       SEARCH suu USING INDEX sqlite_autoindex_unit_1 (uuid=?) LEFT-JOIN       
20      0       0       SEARCH sua USING INDEX sqlite_autoindex_application_1 (uuid=?) LEFT-JOIN
27      0       0       SEARCH scu USING INDEX sqlite_autoindex_unit_1 (uuid=?) LEFT-JOIN       
34      0       0       SEARCH sca USING INDEX sqlite_autoindex_application_1 (uuid=?) LEFT-JOIN
41      0       0       SCAN m USING COVERING INDEX sqlite_autoindex_model_1